Makefile

```diff
@@ -199,46 +230,64 @@ managers: ## Build all managers
 clusterctl: ## Build clusterctl binary
 	go build -trimpath -ldflags "$(LDFLAGS)" -o $(BIN_DIR)/clusterctl sigs.k8s.io/cluster-api/cmd/clusterctl
 
-$(SETUP_ENVTEST): $(TOOLS_DIR)/go.mod # Build setup-envtest from tools folder.
-	cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/setup-envtest sigs.k8s.io/controller-runtime/tools/setup-envtest
+$(SETUP_ENVTEST): # Build setup-envtest from tools folder.
+	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) ${SETUP_ENVTEST_PKG} ${SETUP_ENVTEST_BIN} ${SETUP_ENVTEST_VER}
```

Comment:
Could we have a generic Make target that invokes go install like:

```Makefile
go-install: # Builds a go binary using a temporary go module.
	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) $(GO_PKG) $(GO_OUTPUT_BIN) ${GO_VERSION}
```

And then call it like
```Makefile
$(SETUP_ENVTEST): # Build setup-envtest from tools folder.
   [variables here] $(MAKE) go-install
```

Can we also make sure that when the version for a binary changes, the Makefile/go-install script builds and executes the right one only if necessary?

Something we have today is that tools are only built when needed and if you change the go.mod file, they get rebuilt.

Reply:
> Can we also make sure that when the version for a binary changes, the Makefile/go-install script builds and executes the right one only if necessary?

see my comment below

Reply:
Which one? 🤔

Reply:
I guess above 😅 

https://github.com/kubernetes-sigs/cluster-api/pull/5741#discussion_r759500353

Reply:
This is a bit different than my suggestion above, that's fine if we keep those, although we can probably have most of these binaries builds call the sub-make command `go-install` as shown above

Reply:
> The binary changes should also be dependant on the version though, the current Makefile implementation I think only looks at the name in the tools/bin folder, which means that upon switching branches a different version of a binary wouldn't build the correct one and potentially use a newer/older version unless the user cleans up the binary folder

they do have the version in them currently. For example, `GO_APIDIFF` is `$(abspath $(TOOLS_BIN_DIR)/$(GO_APIDIFF_BIN)-$(GO_APIDIFF_VER))`. It will build the binary if the version in hack/tools/bin doesn't match.

Reply:
That works! I missed that was in the binaries' name